### PR TITLE
pipe-viewer: init at 0.1.4

### PR DESCRIPTION
**Pull request by ju1m**
###### Motivation for this change
Use [`pipe-viewer` and `gtk-pipe-viewer`](https://github.com/trizen/pipe-viewer) to watch Youtube.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 136119` at b2acf478 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>gtk-pipe-viewer</li>
    <li>perl532Packages.LWPUserAgentCached</li>
    <li>perl532Packages.TextParsewords</li>
    <li>perl534Packages.LWPUserAgentCached</li>
    <li>perl534Packages.TextParsewords</li>
    <li>pipe-viewer</li>
  </ul>
</details>

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -12188,6 +12188,19 @@ let
...
+  LWPUserAgentCached = buildPerlPackage {
+    pname = "LWP-UserAgent-Cached";
+    version = "0.08";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/O/OL/OLEG/LWP-UserAgent-Cached-0.08.tar.gz";
+      hash = "sha256-Pc5atMeAQWVs54Vk92Az5b0ew4b1TS57MHQK5I7nh8M=";
+    };
+    propagatedBuildInputs = [ LWP ];
+    meta = {
+      description = "LWP:UserAgent with simple caching mechanism";
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:12200]: Where is the `license`?

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -22206,6 +22219,18 @@ let
...
 
+  TextParsewords = buildPerlPackage {
+    pname = "Text-Parsewords";
+    version = "3.30";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/C/CH/CHORNY/Text-ParseWords-3.30.tar.gz";
+      hash = "sha256-heAjgXndQ5l+WMZr1RYRGCvH1TNQUCmi2w0yMu2v9eg=";
+    };
+    meta = {
+      license = lib.licenses.free; # Same as Perl
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:22230]: If the same as Perl license, then

```suggestion
      license = with lib.licenses; [ artistic1 gpl1Plus ];
```

LWP-UserAgent-Cached also seems to be under this scheme, so this suggestion applies there as well.

**Review by zakame — CHANGES_REQUESTED** [review]
The `perl-packages.nix` changes are worth their own PRs prior to this one - useful on their own right.

File: pkgs/top-level/perl-packages.nix (hunk 2 above)
Comment (zakame) [reply pkgs/top-level/perl-packages.nix:22230]: There should also be a `description` as well.

**Comment by ju1m** [comment]
@zakame, I've applied your suggestions. Thanks.

**Comment by zakame** [comment]
@GrahamcOfBorg build perlPackages.LWPUserAgentCached perlPackages.TextParsewords

File: pkgs/applications/video/pipe-viewer/default.nix [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  src = fetchFromGitHub {
+    owner = "trizen";
+    repo = "pipe-viewer";
+    rev = version;
+    hash = "sha256-XoTOjUzhJD3Q0sENe9m5etqbAfBh83B/xmRmqz8zETU=";
+  };
+
+  nativeBuildInputs = [ makeWrapper ] ++
+    lib.optionals withGtk3 [ wrapGAppsHook ];
```
Comment (SuperSandro2000) [inline pkgs/applications/video/pipe-viewer/default.nix:42]:
```suggestion
  nativeBuildInputs = [ makeWrapper ]
    ++ lib.optionals withGtk3 [ wrapGAppsHook ];
```

File: pkgs/applications/video/pipe-viewer/default.nix [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  checkInputs = [
+    TestPod
+  ];
+
+  postInstall = ''
+    wrapProgram "$out/bin/pipe-viewer" \
+      --prefix PATH : "${lib.makeBinPath [ ffmpeg wget youtube-dl ]}"
+  '' + lib.optionalString withGtk3 ''
+    wrapProgram "$out/bin/gtk-pipe-viewer" \
+      --prefix PATH : "${lib.makeBinPath [ ffmpeg wget xdg-utils youtube-dl ]}"
```
Comment (SuperSandro2000) [inline pkgs/applications/video/pipe-viewer/default.nix:84]: Does this double wrap?

Comment (ju1m) [reply pkgs/applications/video/pipe-viewer/default.nix]: @SuperSandro2000, good catch, it was even a triple wrap for `gtk-pipe-viewer`, I've fixed this by setting `dontWrapGApps` and passing `gappsWrapperArgs` to the `wrapProgram` on `gtk-pipre-viewer`, thanks!

**Comment by ju1m** [comment]
damned, it suddenly rebuilds a lot :o I guess I'll have to rebase against `staging`.

**Comment by ju1m** [comment]
My bad, I had mistakenly included a change to support `buildFlags` in `buildPerlModule`:
https://github.com/NixOS/nixpkgs/pull/136119/files#diff-b8f33964a93f34e19fdee3d032e64396c384b7b652f1e167a708b21ab79a0ba9L53-R53
It should return back to a more reasonable build now.

File: pkgs/applications/video/pipe-viewer/default.nix [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+, JSONXS
+, LWPProtocolHttps
+, LWPUserAgentCached
+, Memoize
+, PathTools
+, ScalarListUtils
+, TermReadLineGnu
+, TestPod
+, TextParsewords
+, UnicodeLineBreak
```
Comment (zakame) [inline pkgs/applications/video/pipe-viewer/default.nix:27]: Maybe we can use a `perl.withPackages()` environment for this?  Is there a requirement for pipe-viewer to require these Perl packages directly on the user environment?

File: pkgs/top-level/perl-packages.nix [hunk 6]
```diff
@@ -12188,6 +12188,20 @@ let
...
+  LWPUserAgentCached = buildPerlPackage {
+    pname = "LWP-UserAgent-Cached";
+    version = "0.08";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/O/OL/OLEG/LWP-UserAgent-Cached-0.08.tar.gz";
+      hash = "sha256-Pc5atMeAQWVs54Vk92Az5b0ew4b1TS57MHQK5I7nh8M=";
+    };
+    propagatedBuildInputs = [ LWP ];
+    meta = {
+      description = "LWP:UserAgent with simple caching mechanism";
```
Comment (zakame) [inline pkgs/top-level/perl-packages.nix:12200]:
```suggestion
      description = "LWP::UserAgent with simple caching mechanism";
```

The `::` is significant; alternatively we can also use

```suggestion
      description = "LWP-UserAgent with simple caching mechanism";
```

**Review by zakame — CHANGES_REQUESTED** [review]
Thanks, we're getting close! 🙏 I'd suggest using a `perlEnv` pattern to keep all Perl dependencies in their own derivation/environment if at all possible, something like

```nix
{ perl
, buildPerlModule
, withGtk3 ? false 
, ...
}:
let
  perlEnv = perl.withPackages(ps: with ps; [
    AnyURIEscape
    DataDump
    Encode
    ...
  ] ++ lib.optionals withGtk3 [ FileShareDir Gtk3 ]);
in
buildPerlModule rec {
  ...
  buildInputs = [ perlEnv ];
  ...
}
```

File: pkgs/applications/video/pipe-viewer/default.nix (hunk 5 above)
Comment (ju1m) [reply pkgs/applications/video/pipe-viewer/default.nix]: @zakame, I've done that, except for `Gtk3` because it fails `wrapGAppsHook`.

**Pull request merged**